### PR TITLE
docs: add integration testing guide

### PR DESCRIPTION
**Pull request by corymhall**
This adds a new integration testing guide to help contributors better
understand how CDK integration testing works and under what scenarios
integration tests should be added. This is not meant to be a
comprehensive guide and can serve as a starting point for us to add
additional scenarios, etc.

This also adds a checklist item to the pull request template for
contributors to indicate that their new feature contains a corresponding
integration test and that they have followed the integration test guide.

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](../CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* ~[ ] This PR adds new unconventional dependencies following the process described [here](../CONTRIBUTING.md/#adding-new-unconventional-dependencies)~

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -9,4 +9,9 @@
 
 * [ ] This PR adds new unconventional dependencies following the process described [here](../CONTRIBUTING.md/#adding-new-unconventional-dependencies)
 
+### New Features
+
+* [ ] Have you added the new feature to an [integration test](../INTEGRATION_TESTS.md)?
+	* [ ] Did you use `cdk-integ` to deploy the infrastructure and generate the snapshot (i.e. `cdk-integ --no-dry-run`)?
```
Comment (rix0rrr) [inline .github/PULL_REQUEST_TEMPLATE.md:15]: We don't have to pass `--no-dry-run`, just not passing the flag is good enough.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -9,4 +9,9 @@
 
 * [ ] This PR adds new unconventional dependencies following the process described [here](../CONTRIBUTING.md/#adding-new-unconventional-dependencies)
 
+### New Features
+
+* [ ] Have you added the new feature to an [integration test](../INTEGRATION_TESTS.md)?
```
Comment (rix0rrr) [inline .github/PULL_REQUEST_TEMPLATE.md:14]: I'm not convinced this link is going to work when viewed in a pull request...?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -234,9 +234,17 @@ Integration tests perform a few functions in the CDK code base -
 3. (Optionally) Acts as a way to validate that constructs set up the CloudFormation resources as expected. A successful
    CloudFormation deployment does not mean that the resources are set up correctly.
 
-If you are working on a new feature that is using previously unused CloudFormation resource types, or involves
-configuring resource types across services, you need to write integration tests that use these resource types or
-features.
+**When are integration tests required?**
+
+The following list contains common scenarios where we _know_ that integration tests are required.
+This is not an exhaustive list and we will, by default, require integration tests for all
```
Comment (otaviomacedo) [inline CONTRIBUTING.md:240]: Maybe also when adding a new Context Provider?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (corymhall) [reply .github/PULL_REQUEST_TEMPLATE.md]: Updated.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (corymhall) [reply .github/PULL_REQUEST_TEMPLATE.md]: yeah you are right. Updated all the links in the template.

File: CONTRIBUTING.md (hunk 3 above)
Comment (corymhall) [reply CONTRIBUTING.md:240]: Where do those integration tests live today?

**Review by otaviomacedo — APPROVED** [review]

**Pull request merged**